WordPress/gutenberg
# Gallery: Add lightbox support

Originally written by @madhusudhand 

<details><summary>Details</summary>

## What?

This introduces lightbox functionality to `Gallery` block.

## How?

* Image lightbox has been extended to support gallery
* Next and Prev icons were added to image lightbox if the image is in the context of a gallery.
* In this iteration (first) next/prev navigation for the lightbox in gallery are based on the "Expand on click" property of the image. 

Following is the behaviour in various scenarios.

### Case 1: "Expand on click" is enabled for the image block globally.

All images inside (unless overridden at block level), will invoke the lightbox on click and next and previous navigation is possible.

### Case 2: One of the image (lets say 2nd image) is set to "Link to image file"

Lightbox will be invoked on click of 1st image and on click of next will move to 3rd image.

### Case 3: "Expand on click" is **not** enabled globally, but few images are set with this option in gallery.

This is similar to **case 2**, lightbox opens with only those images.

## What's in future PRs?

* This change updates the image source on next and prev navigation. The smooth transition between the images will be addressed in a future PR.
* This has no effect on touch devices. It will be addressed in followup.

## Screenshots (Draft)

![gallery-lightbox-1](https://github.com/user-attachments/assets/75f237b8-8db4-45a9-b7d6-f3c617060686)

<img width="636" alt="image" src="https://github.com/user-attachments/assets/d6fa7f79-3104-4895-a7d8-10e344a3cc62">

## Testing instructions

* In global styles, enable "Click to expand" feature on image block.
* Add a gallery block to a post or page. Add images to it.
* In the frontend, click on an image to invoke lightbox and navigate between images by clicking on icons in the UI, or using next and prev arrow keys in the keyboard.

# Related issues

First iteration of #56310

</details> 

---

Latest description written by @t-hamano


## What?

This PR enables navigation between images with lightbox enabled.

## Why?

#56310

## How?

- If the Gallery block contains images with "Enlarge on click" enabled, you will be able to navigate between those images. If "Enlarge on click" is not enabled, those images will be skipped.
- This feature is enabled **by default**, ~but can be disabled in the block sidebar. If the feature is disabled, the image lightboxes themselves will work as before, but navigation between images will not be enabled:~ Update: Changed to always be enabled.
- Swipe actions on touch devices are not currently possible. This may be addressed in a follow-up.
- To localize the `aria-label`, inject an element with `aria-live` and `aria-atomic` attributes and vary the text inside. We could also use `speak`, but [the `@wordpress/a11y` script module is available from WP 6.7](https://make.wordpress.org/core/2024/10/14/updates-to-script-modules-in-6-7/). The Gutenberg plugin has to support WP 6.6 for now, so I didn't go with the speak approach.

## Testing Instructions

- Click the "Enlarge on click" button in the Gallery block toolbar. This will enable Lightbox for all images.
- When you click on an image, a lightbox will open. Make sure the close, back, and next buttons work.
- ~Disable "Lightbox navigation" in the Gallery block sidebar. A lightbox will open when you click on an image, but the back and next buttons will not be displayed.~

### Testing Instructions for Keyboard

- Make sure you can perform all test steps using only the keyboard.
- ~Tooltip text appears when the back and next buttons are hovered over or focused.~ Update: Based on [this feedback](https://github.com/WordPress/gutenberg/pull/62906#issuecomment-2678565124), we removed the tooltip and instead added the aria-label attribute to the button.
- Make sure your screen reader reads the appropriate information. For example, NVDA reads:
  - Click lightbox toggle button: `Enlarged image X of Y: {IMAGE_ALT_TEXT} dialog`
  - Focus back and next buttons: `button unavailable Previous`, `button Next`
  - Click back and next buttons: `Enlarged image X of Y: {IMAGE_ALT_TEXT}`

## Screenshots or screencast

https://github.com/user-attachments/assets/4e6d3a13-d0a4-472c-b654-b8909b4e9788
